src/Features/CSharp/Portable/CodeMapper/CSharpMapCodeService.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeCleanup;
+using Microsoft.CodeAnalysis.CodeMapping;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeMapping;
+
+/// <summary>
+/// A mapper used for Mapping C# Code from a suggestion onto existing code in the codebase.
+/// This is the backbone of previews, as this allows us to know where we should place the preview.
+/// </summary>
+[ExportLanguageService(typeof(IMapCodeService), language: LanguageNames.CSharp), Shared]
+internal sealed partial class CSharpMapCodeService : IMapCodeService
+{
+    private readonly IGlobalOptionService _globalOptions;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public CSharpMapCodeService(IGlobalOptionService globalOptions)
+    {
+        _globalOptions = globalOptions;
+    }
+
+    public async Task<Document> MapCodeAsync(Document document, ImmutableArray<string> contents, ImmutableArray<DocumentSpan> focusLocations, bool formatMappedCode, CancellationToken cancellationToken)
+    {
+        var target = await GetMappingTargetAsync(document, focusLocations, cancellationToken).ConfigureAwait(false);
```

Comment:
I was expecting this (GetMappingTargetAsync) to remain on the VS Side, what's the reasoning behind including this here?

Reply:
oh wait, I see it only has the same name, but it does different things.

Reply:
I'd suggest to review this commit by commit. I tried to group relevant changes into separate commits